File: backend/api/app/models/tracking_event.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, String, Text, func
from app.database import GUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class TrackingEvent(Base):
    __tablename__ = "tracking_events"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    shipment_id: Mapped[uuid.UUID] = mapped_column(GUID(), ForeignKey("shipments.id"), nullable=False, index=True)
    event_type: Mapped[str] = mapped_column(String(100), nullable=False)
    new_status: Mapped[str | None] = mapped_column(String(50), nullable=True)
    location: Mapped[str | None] = mapped_column(String(255), nullable=True)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    user_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("users.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), index=True)

    shipment = relationship("Shipment", back_populates="tracking_events")

File: packages/shared/lib/src/models/tracking_event.dart
import '../enums/shipment_status.dart';
import '../enums/tracking_event_type.dart';

class TrackingEvent {
  final String id;
  final String shipmentId;
  final TrackingEventType eventType;
  final ShipmentStatus? newStatus;
  final String? location;
  final double? latitude;
  final double? longitude;
  final String? description;
  final String? userId;
  final DateTime createdAt;

  TrackingEvent({
    required this.id,
    required this.shipmentId,
    required this.eventType,
    this.newStatus,
    this.location,
    this.latitude,
    this.longitude,
    this.description,
    this.userId,
    required this.createdAt,
  });

  factory TrackingEvent.fromJson(Map<String, dynamic> json) {
    return TrackingEvent(
      id: json['id'] as String,
      shipmentId: json['shipment_id'] as String,
      eventType: TrackingEventType.fromApi(json['event_type'] as String),
      newStatus: json['new_status'] != null
          ? ShipmentStatus.fromApi(json['new_status'] as String)
          : null,
      location: json['location'] as String?,
      latitude: (json['latitude'] as num?)?.toDouble(),
      longitude: (json['longitude'] as num?)?.toDouble(),
      description: json['description'] as String?,
      userId: json['user_id'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'shipment_id': shipmentId,
      'event_type': eventType.apiValue,
      'new_status': newStatus?.apiValue,
      'location': location,
      'latitude': latitude,
      'longitude': longitude,
      'description': description,
      'user_id': userId,
      'created_at': createdAt.toIso8601String(),
    };
  }
}
